homeassistant/components/overkiz/executor.py

```diff
@@ -0,0 +1,141 @@
+"""Class for helpers and community with the OverKiz API."""
+from __future__ import annotations
+
+import logging
+from typing import Any
+from urllib.parse import urlparse
+
+from pyoverkiz.models import Command, Device
+
+from .coordinator import OverkizDataUpdateCoordinator
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class OverkizExecutor:
+    """Representation of an Overkiz device with execution handler."""
+
+    def __init__(self, device_url: str, coordinator: OverkizDataUpdateCoordinator):
+        """Initialize the executor."""
+        self.device_url = device_url
+        self.coordinator = coordinator
+        self.base_device_url, *_ = self.device_url.split("#")
+
+    @property
+    def device(self) -> Device:
+        """Return Overkiz device linked to this entity."""
+        return self.coordinator.data[self.device_url]
+
+    def select_command(self, *commands: str) -> str | None:
+        """Select first existing command in a list of commands."""
+        existing_commands = self.device.definition.commands
+        return next((c for c in commands if c in existing_commands), None)
+
+    def has_command(self, *commands: str) -> bool:
+        """Return True if a command exists in a list of commands."""
+        return self.select_command(*commands) is not None
+
+    def select_state(self, *states) -> str | None:
+        """Select first existing active state in a list of states."""
+        if self.device.states:
+            return next(
+                (
+                    state.value
+                    for state in self.device.states
+                    if state.name in list(states)
+                ),
+                None,
+            )
+        return None
+
+    def has_state(self, *states: str) -> bool:
+        """Return True if a state exists in self."""
+        return self.select_state(*states) is not None
+
+    def select_attribute(self, *attributes) -> str | None:
+        """Select first existing active state in a list of states."""
+        if self.device.attributes:
+            return next(
+                (
+                    attribute.value
+                    for attribute in self.device.attributes
+                    if attribute.name in list(attributes)
+                ),
+                None,
+            )
```

Comment:
Can this linear search be avoided?

Reply:
What about:

```Python
def select_attribute(self, *attributes) -> Optional[str]:
    """Select first existing active state in a list of states."""
    for attribute in attributes:
        if current_attribute := self.device.attributes[attribute]:
            return current_attribute.value
    return None
```

There is always an attributes property, so we can remove the 1st condition.

Reply:
Applied your new code statement, however it is still some kind of linear search, just a bit more optimized..

Reply:
Ideally this gets optimized in the future, but its ok for now